src/main/scala/org/broadinstitute/dsde/workbench/sam/util/ShadowCaster.scala

```diff
@@ -0,0 +1,39 @@
+package org.broadinstitute.dsde.workbench.sam.util
+import cats.effect._
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.concurrent.duration.MILLISECONDS
+import scala.concurrent.{ExecutionContext, Future}
+
+trait ShadowCaster extends LazyLogging {
+  implicit val executionContext: ExecutionContext
+
+  protected def runWithShadow[T](functionName: String, real: IO[T], shadow: IO[T]): IO[T] = {
+
+    val realFuture = measure(real).unsafeToFuture()
+    val shadowFuture = measure(shadow).unsafeToFuture()
+
+    Future.reduceLeft[(T, Long), (T, Long)](List(realFuture, shadowFuture)) { case ((realResult, realTime), (shadowResult, shadowTime)) =>
+      reportResult(functionName, realResult, realTime, shadowResult, shadowTime)
+      (realResult, realTime)
+    }
+
+    IO.fromFuture(IO(realFuture)).map(_._1)
+  }
+
+  protected def reportResult[T](functionName: String, realResult: T, realTime: Long, shadowResult: T, shadowTime: Long) = {
+    val eq = realResult != shadowResult
+    val timeDiff = shadowTime - realTime
+
+    logger.debug(s"$functionName, equal: $eq, time increase $timeDiff")
```

Comment:
is it useful to say _how_ they're not equal here, for easier debugging? scalatest writes fairly good output for `shouldEqual`, maybe you can steal it

Reply:
we will spend more time figuring out how to report